### DISCUSSION TITLE
Xero Tracking Categories

### DISCUSSION THREAD

**Discussion by AndyTurff** [XeroAPI/xero-php-oauth2#315] [Q&A] [answered]
Hi, I'm struggling with sending Purchase Orders with 2 tracking categories on a line item. I'm currently sending them with with one tracking category  but I can't work out the syntax for send two different tracking categories. This is what I use for sending one...

$lineItemTracking[$x] = new XeroAPI\XeroPHP\Models\Accounting\LineItemTracking;
	$lineItemTracking[$x]
			->setName('Dept')
			->setOption($DeptX);
			array_push($lineItemTrackings, $lineItemTracking[$x]);

So what I need to achieve is...

$lineItemTracking[$x] = new XeroAPI\XeroPHP\Models\Accounting\LineItemTracking;
	$lineItemTracking[$x]
			->setName('Dept')
			->setOption($myDeptX);
			->setName('Unit')
			->setOption($myUnitX);
			array_push($lineItemTrackings, $lineItemTracking[$x]);

Hope someone can help. It would be very much appreciated.

**Comment by pumpkinball** [comment]
Hi Andy,
Off the top of my head, have you tried instantiating the LineItemTracking object again?
So for each LineItemTracking you would need to do 

$lineItemTracking[1] = new XeroAPI\XeroPHP\Models\Accounting\LineItemTracking;
blah blah ... 
$lineItemTracking[2] = new XeroAPI\XeroPHP\Models\Accounting\LineItemTracking;
blah blah
And then at the end 
array_push($lineItemTrackings, $lineItemTracking);

**Comment by AndyTurff** [comment]
Hi pumpkinball.

Thank you for your quick reply. I tried something similar which didn't work
but I'll give what you suggest in the morning. Thank you again.

Andy

**Reply by AndyTurff** [reply]
So I tried as listed below (the xxxx's are mine) 

	$lineitemsX[$x] = new XeroAPI\XeroPHP\Models\Accounting\LineItem;
	$lineItemTracking[1] = new XeroAPI\XeroPHP\Models\Accounting\LineItemTracking;
		$lineItemTracking[1]
			->setTrackingCategoryID('xxxxxxxx-xxxx-xxxx-947d-0234e5de3833')
			->setTrackingOptionID('xxxxxxxx-xxxx-xxxx-b847-c55637fef858')
			->setName('Dept')
			->setOption('09 - General');

	$lineItemTracking[2] = new XeroAPI\XeroPHP\Models\Accounting\LineItemTracking;
		$lineItemTracking[2]
			->setTrackingCategoryID('xxxxxxxx-xxxx-xxxx-21630393aea2')
			->setTrackingOptionID('xxxxxxxx-xxxx-xxxx-b4bb-860c184f3b84')
			->setName('Unit')
			->setOption('All Units');
	array_push($lineItemTrackings, $lineItemTracking);

		$lineitemsX[$x]
			->setDescription($partNumX[$x])
			->setQuantity($qtyX[$x])
			->setUnitAmount($unitX[$x])
			->setAccountCode($codeX[$x])
			->setDiscountRate($discountX[$x])
			->setTaxType($taxXero[$x])
			->setTracking($lineItemTrackings);
			array_push($lineitems, $lineitemsX[$x]);

This a print_r of the relevant part of my $PurchaseOrder, which gets no syntax errors but when I print_r the $result I get the bottom 2 paragraphs. So it seems to not be sending the value through and I'm still only getting 1 set of tracking category. I'm sure I must still be getting the syntax wrong but I can't work out the solution.

 [tracking] => Array ( [0] => Array ( [1] => XeroAPI\XeroPHP\Models\Accounting\LineItemTracking Object ( [container:protected] => Array ( [tracking_category_id] => xxxxxxxx-xxxx-xxxx-947d-0234e5de3833 [tracking_option_id] => 9a4bbdf9-ca2d-4427-b847-c55637fef858 [name] => Dept [option] => 14 - SC - General ) ) [2] => XeroAPI\XeroPHP\Models\Accounting\LineItemTracking Object ( [container:protected] => Array ( [tracking_category_id] => xxxxxxxx-eddc-xxxx-xxxx-21630393aea2 [tracking_option_id] => xxxxxxxx-xxxx-xxxx-xxxx-860c184f3b84 [name] => Unit [option] => All Units ) ) ) )
 
 [tracking] => Array ( [0] => XeroAPI\XeroPHP\Models\Accounting\LineItemTracking Object ( [container:protected] => Array ( [tracking_category_id] => 00000000-0000-0000-0000-000000000000 [tracking_option_id] => [name] => [option] => ) ) )
 
 ERROR [validation_errors] => Array ( [0] => XeroAPI\XeroPHP\Models\Accounting\ValidationError Object ( [container:protected] => Array ( [message] => Unknown tracking details. A Tracking Category did not match a known Active TrackingCategory ) ) )

Any ideas?

Thank you

**Reply by AndyTurff** [reply]
I also tried this, with a similar result...

 	/*Standard line items*/
	$lineitemsX[$x] = new XeroAPI\XeroPHP\Models\Accounting\LineItem;
	
	$lineItemTracking[1] = new XeroAPI\XeroPHP\Models\Accounting\LineItemTracking;
		$lineItemTracking[1]
			->setName('Dept')
			->setOption($DeptX[$x]);

		$lineItemTracking[2]
			->setName('Unit')
			->setOption($xeroUnit[$x]);
	array_push($lineItemTrackings, $lineItemTracking);
 

 [tracking] => Array ( [0] => Array ( [1] => XeroAPI\XeroPHP\Models\Accounting\LineItemTracking Object ( [container:protected] => Array ( [tracking_category_id] => xxxxxxxx-xxxx-xxxx-947d-0234e5de3833 [tracking_option_id] => xxxxxxxx-xxxx-xxxx-b847-c55637fef858 [name] => Dept [option] => 14 - SC - General ) ) [2] => XeroAPI\XeroPHP\Models\Accounting\LineItemTracking Object ( [container:protected] => Array ( [tracking_category_id] => xxxxxxxx-xxxx-xxxx-21630393aea2 [tracking_option_id] => xxxxxxxx-xxxx-xxxx-b4bb-860c184f3b84 [name] => Unit [option] => All Units ) ) ) )
 
 
 [tracking] => Array ( [0] => XeroAPI\XeroPHP\Models\Accounting\LineItemTracking Object ( [container:protected] => Array ( [tracking_category_id] => 00000000-0000-0000-0000-000000000000 [tracking_option_id] => [name] => [option] => ) ) )

**Comment by pumpkinball** [comment]
Hi @AndyTurff I know it seems a daft question, but the error suggests that the tracking categories you are sending across are not active.. ? 
If you give me you clientID, I can look in the logs for you.

**Reply by AndyTurff** [reply]
Hi, thanks for your continued help. I know its not that because if I send a purchase order through with either just Unit or Dept it works fine just not with them both.

the $result looks like this when there is just one...

 [tracking] => Array ( [0] => XeroAPI\XeroPHP\Models\Accounting\LineItemTracking Object ( [container:protected] => Array ( [tracking_category_id] => xxxxxxxx-xxxx-xxxx-af54-21630393aea2 [tracking_option_id] => xxxxxxxx-xxxx-xxxx-b4bb-860c184f3b84 [name] => Unit [option] => All Units ) ) )

And like this when there is 2 categories (it does't even show the second array)

[tracking] => Array ( [0] => XeroAPI\XeroPHP\Models\Accounting\LineItemTracking Object ( [container:protected] => Array ( [tracking_category_id] => 00000000-0000-0000-0000-000000000000 [tracking_option_id] => [name] => [option] => ) ) ) 

All the variables are mising and I think that's why it says "A Tracking Category did not match a known Active TrackingCategory"

I still think it's because the syntax is wrong but I've tried so many different ways and I am stumped!

**Comment by pumpkinball** [comment, marked as answer]
Hi @AndyTurff I've just tried it on my local. With the array_push, I forgot you have to push into the array after each tracking item. 
So I tried this, and it seemed to work.. 
    
```
             $trackingCat = new XeroAPI\XeroPHP\Models\Accounting\LineItemTracking;
	    $trackingCat->setTrackingCategoryId('xxxxxxxx-xxxx-xxxx-97a1-0e6d15b0b574')
					   ->setName('Garden')
				            ->setOption('All Gardens');

		array_push($trackingCats, $trackingCat);	

	    $trackingCat = new XeroAPI\XeroPHP\Models\Accounting\LineItemTracking;
	    $trackingCat->setTrackingCategoryId('xxxxxxxx-xxxx-xxxx-8dcb-9405d571b614')
					   ->setName('Unit')
				            ->setOption('All Units');

	    array_push($trackingCats, $trackingCat);	
	
		$lineitems = [];		

		$lineitem = new XeroAPI\XeroPHP\Models\Accounting\LineItem;
		$lineitem->setDescription('Item 123')
			->setQuantity(1)
			->setUnitAmount(32)
			->setAccountCode("400")
			->setTracking($trackingCats);

		array_push($lineitems, $lineitem);
```

**Comment by AndyTurff** [comment]
That worked!

Thank you so much for all your help. It's very much appreciated.
Have a great weekend. I will now! 🤓

Andy

**Reply by pumpkinball** [reply]
Good to hear! Enjoy!!
